Formula/gobject-introspection-utils.rb

```diff
@@ -0,0 +1,101 @@
+  url "https://download.gnome.org/sources/gobject-introspection/1.72/gobject-introspection-1.72.0.tar.xz"
+  sha256 "02fe8e590861d88f83060dd39cda5ccaa60b2da1d21d0f95499301b186beaabc"
+  license all_of: ["GPL-2.0-or-later", "LGPL-2.0-or-later", "MIT"]
+
+  depends_on "bison" => :build
+  depends_on "meson" => :build
+  depends_on "ninja" => :build
+  depends_on "cairo"
+  depends_on "glib"
+  depends_on "gobject-introspection"
+  depends_on "libffi"
+  depends_on "pkg-config"
+  depends_on "python@3.10"
+
+  uses_from_macos "flex" => :build
+
+  def install
+    ENV["GI_SCANNER_DISABLE_CACHE"] = "true"
+    inreplace "giscanner/transformer.py", "/usr/share", "#{HOMEBREW_PREFIX}/share"
+    inreplace "meson.build",
+      "config.set_quoted('GOBJECT_INTROSPECTION_LIBDIR', join_paths(get_option('prefix'), get_option('libdir')))",
+      "config.set_quoted('GOBJECT_INTROSPECTION_LIBDIR', '#{HOMEBREW_PREFIX}/lib')"
+
+    # Point the headers and libraries to `gobject-introspection`.
+    # The headers and libraries will be removed later because they are provided by `gobject-introspection`.
+    gobject_introspection = Formula["gobject-introspection"]
+    args = std_meson_args.delete_if do |arg|
+      arg.start_with?("--includedir=", "--libdir=")
+    end
+    args += %W[
+      --includedir=#{gobject_introspection.opt_include}
+      --libdir=#{gobject_introspection.opt_lib}
+    ]
+
+    mkdir "build" do
+      system "meson", *args,
+        "-Dpython=#{Formula["python@3.10"].opt_bin}/python3",
+        ".."
+      system "ninja", "-v"
+
+      # Skip files already in gobject-introspection.
+      Formula["meson"].opt_libexec.cd do
+        system "bin/python3", "-c", pyscript
+      end
+      system "ninja", "install", "-v"
+    end
+
+    # Delete non python files because they are provided by `gobject-introspection`
+    python_extension_regex = /\.(py(?:[diwx])?|px[di]|cpython-(?:[23]\d{1,2})[-\w]*\.(so|dylib))$/i
+    python_shebang_regex = %r{^#! ?/usr/bin/(?:env )?python(?:[23](?:\.\d{1,2})?)?( |$)}
+    shebang_max_length = 28 # the length of "#! /usr/bin/env pythonx.yyy "
+    prefix.find do |f|
+      next unless f.file?
+      next if python_extension_regex.match?(f.basename) || python_shebang_regex.match?(f.read(shebang_max_length))
+
+      f.unlink
+    end
+
+    # Delete empty directories
+    # Note: We need to traversal the directories in reverse order (i.e. deepest first).
+    #       Also, we should put checking emptiness and deletion in a single loop.
+    #       `dirs.select(&:empty?).map(&:rmdir)` will not work because it will not delete
+    #       directories that only contain empty directories.
+    prefix.find.select(&:directory?).reverse_each { |d| d.rmdir if d.empty? }
+
+    # Install giscanner Python library.
+    (lib/"gobject-introspection/giscanner").install (buildpath/"build/giscanner").glob(["*.py", "*.so", "*.a"])
+
+    rewrite_shebang detected_python_shebang, *bin.children
+  end
+
+  def pyscript
```

Comment:
If we're using this in multiple formulae we should probably handle this in `brew`.

Reply:
As before, I'm really not happy about messing with `meson` internals here.

Reply:
> As before, I'm really not happy about messing with `meson` internals here.

This is the only way we can avoid accessing the other formula's cellar, because we use `--includedir=#{otherformula.opt_include}` and `--libdir=#{otherformula.opt_lib}` in `meson` configuration arguments

Unless we can let the formula have the write access to another's cellar (e.g. `glib-utils` write to `opt/glib` which is symlinked to `Cellar/glib/<version>`). This is not allowed by the `sandbox` mechanism. Use `glib` and `glib-utils` for example, to avoid messing with `meson` internals, we need to:

1. unlink the `glib` opt prefix `$HOMEBREW_PREFIX/opt/glib`.
2. `ninja install`: install files of `glib-utils` into `$HOMEBREW_PREFIX/opt/glib`.
3. remove everything in `$HOMEBREW_PREFIX/opt/glib`.
4. link `glib` opt prefix back `$HOMEBREW_CELLAR/glib/<version>` to `$HOMEBREW_PREFIX/opt/glib`.

------

To avoid modifying the `meson` internal install data (which structure may change over releases), we may need to ask the `meson` upstream to implement a new feature like:

```bash
meson install --exclude <pattern>
```

Note that this only works for file names, sometimes we need to read the file content for shebang detection.

```console
$ meson install --help
usage: meson install [-h] [-C WD] [--no-rebuild] [--only-changed] [--quiet] [--destdir DESTDIR] [--dry-run] [--skip-subprojects [SKIP_SUBPROJECTS]] [--tags TAGS] [--strip]

options:
  -h, --help                             show this help message and exit
  -C WD                                  directory to cd into before running
  --no-rebuild                           Do not rebuild before installing.
  --only-changed                         Only overwrite files that are older than the copied file.
  --quiet                                Do not print every file that was installed.
  --destdir DESTDIR                      Sets or overrides DESTDIR environment. (Since 0.57.0)
  --dry-run, -n                          Doesn't actually install, but print logs. (Since 0.57.0)
  --skip-subprojects [SKIP_SUBPROJECTS]  Do not install files from given subprojects. (Since 0.58.0)
  --tags TAGS                            Install only targets having one of the given tags. (Since 0.60.0)
  --strip                                Strip targets even if strip option was not set during configure. (Since 0.62.0)
```

Reply:
Alternatively, we could add support to `brew` for building multiple bottles/formulae from a single build.